bin/remote-prover/src/utils.rs

```diff
@@ -1,52 +1,124 @@
 use std::net::TcpListener;
 
+use http::HeaderMap;
 use miden_remote_prover::error::RemoteProverError;
 use pingora::{Error, ErrorType, http::ResponseHeader, protocols::http::ServerSession};
 use pingora_proxy::Session;
+use prost::Message;
+use tonic::Code;
 use tracing::debug;
 
 use crate::{COMPONENT, commands::PROXY_HOST, proxy::metrics::QUEUE_DROP_COUNT};
 
-const RESOURCE_EXHAUSTED_CODE: u16 = 8;
+/// Write a protobuf message as a gRPC response to a Pingora session
+///
+/// This helper function takes a protobuf message and writes it to a Pingora session
+/// in the proper gRPC format, handling message encoding, headers, and trailers.
+pub async fn write_grpc_response_to_session<T>(
+    session: &mut Session,
+    message: T,
+    grpc_status: Option<&str>,
+) -> pingora_core::Result<bool>
+where
+    T: Message,
+{
+    // Serialize the protobuf message
+    let mut response_body = Vec::new();
+    message.encode(&mut response_body).map_err(|e| {
+        Error::new(ErrorType::InternalError)
+            .more_context(format!("Failed to encode proto response: {e}"))
+    })?;
 
-/// Create a 503 response for a full queue
-pub(crate) async fn create_queue_full_response(
+    let mut grpc_message = Vec::new();
+
+    // Add compression flag (1 byte, 0 = no compression)
+    grpc_message.push(0u8);
+
+    // Add message length (4 bytes, big-endian)
+    let msg_len = response_body.len() as u32;
+    grpc_message.extend_from_slice(&msg_len.to_be_bytes());
+
+    // Add the actual message
+    grpc_message.extend_from_slice(&response_body);
+
+    // Create gRPC response headers WITHOUT grpc-status (that goes in trailers)
+    let mut header = ResponseHeader::build(200, None)?;
+    header.insert_header("content-type", "application/grpc".to_string())?;
```

Comment:
Also elsewhere similarly (move the constant to somewhere appropriate I guess).

```suggestion
    const GRPC_CONTENT: HeaderValue = HeaderValue::from_static("application/grpc");
    header.insert_header(http::header::CONTENT_TYPE, GRPC_CONTENT).unwrap();
```